core/node/node_framework/src/service/tests.rs

```diff
@@ -0,0 +1,235 @@
+    async fn run(self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        let mut guard = self.0.lock().await;
+        *guard = true;
+        Ok(())
+    }
+}
+
+// Layer for task that updates resource and returns only after stop signal received
+#[derive(Debug)]
+struct RemainingTaskLayer(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl WiringLayer for RemainingTaskLayer {
+    fn layer_name(&self) -> &'static str {
+        "remaining_task_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut node: ServiceContext<'_>) -> Result<(), WiringError> {
+        node.add_task(Box::new(RemainingTask(self.0.clone())));
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct RemainingTask(Arc<Mutex<bool>>);
+
+#[async_trait::async_trait]
+impl Task for RemainingTask {
+    fn name(&self) -> &'static str {
+        "remaining_task"
+    }
+    async fn run(self: Box<Self>, stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.0.borrow() {
+                let mut guard = self.0.lock().await;
+                *guard = false;
+                break;
+            }
+        }
+        Ok(())
+    }
+}
+
+// Check ZkStack Service's `run()` method expected behavior.
+// There are 2 tests, one for successful task run and another for remaining task run.
+// Second test case relies on the first one.
+#[test]
+fn test_task_run() {
+    // case 1: run the task and check it updates resource successfully before stop signal received.
+    let resource = Arc::new(Mutex::new(false));
+    let mut zk_stack_service = ZkStackService::new().unwrap();
+    zk_stack_service.add_layer(SuccessfulTaskLayer(resource.clone()));
+
+    assert!(zk_stack_service.run().is_ok());
+    // task changed resource value to true
+    tokio::runtime::Runtime::new()
+        .unwrap()
+        .block_on(async { assert!(*resource.lock().await) });
+
+    // case 2: the idea is to check we are still running remaining tasks after stop signal received.
+    // The flow is that successful_task updates resource to true (as checked in the case 1 above)
+    // and remaining_task updates it back to false.
+    let resource = Arc::new(Mutex::new(false));
+    let mut zk_stack_service = ZkStackService::new().unwrap();
+    zk_stack_service
+        .add_layer(SuccessfulTaskLayer(resource.clone()))
+        .add_layer(RemainingTaskLayer(resource.clone()));
+
+    assert!(zk_stack_service.run().is_ok());
+    tokio::runtime::Runtime::new()
+        .unwrap()
+        .block_on(async { assert!(!*resource.lock().await) });
+}
+
+// Check ZkStack Service's `stop_receiver()` method returns new `Receiver` for it's own sender,
+// and not arbitrary one.
+#[test]
+fn test_stop_receiver() {
+    let zk_stack_service = ZkStackService::new().unwrap();
+    let _receiver1 = zk_stack_service.stop_sender.subscribe();
+    let _receiver2 = zk_stack_service.stop_receiver();
+    assert_eq!(2, zk_stack_service.stop_sender.receiver_count());
+}
```

Comment:
I'd argue that it's an implementation detail. You have a test that checks if the stop receiver works in general -- it's sufficient. The implementation may change, and gluing tests to some specific implementation makes development unnecessarily hard.

Reply:
Do you suggest getting rid of this?

Reply:
Yup